pkg/controllers/state/metrics/node.go

```diff
@@ -0,0 +1,251 @@
+	ns.cluster.ForEachNode(func(n *state.Node) bool {
+		if _, ok := ns.labelMap[n.Node.Name]; !ok {
+			logging.FromContext(ctx).Infof("Tracking new node: %s", n.Node.Name)
+			ns.labelMap[n.Node.Name] = make(map[*prometheus.GaugeVec][]prometheus.Labels)
+		}
+		nodes[n.Node.Name] = struct{}{}
+
+		podRequests := resources.Subtract(n.PodTotalRequests, n.DaemonSetRequested)
+		podLimits := resources.Subtract(n.PodTotalLimits, n.DaemonSetLimits)
+		allocatable := n.Node.Status.Capacity
+		if len(n.Node.Status.Allocatable) > 0 {
+			allocatable = n.Node.Status.Allocatable
+		}
+
+		// Populate  metrics
+		for gaugeVec, resourceList := range map[*prometheus.GaugeVec]v1.ResourceList{
+			overheadGaugeVec:       ns.getSystemOverhead(n.Node),
+			podRequestsGaugeVec:    podRequests,
+			podLimitsGaugeVec:      podLimits,
+			daemonRequestsGaugeVec: n.DaemonSetRequested,
+			daemonLimitsGaugeVec:   n.DaemonSetLimits,
+			allocatableGaugeVec:    allocatable,
+		} {
+			if err := ns.set(resourceList, n.Node, gaugeVec); err != nil {
+				logging.FromContext(ctx).Errorf("Failed to generate gauge: %s", err)
+			}
+		}
+
+		return true
+	})
+
+	ns.cleanup(ctx, nodes)
+}
+
+func (ns *nodeScraper) reset() {
+	for _, gauge := range []*prometheus.GaugeVec{
+		allocatableGaugeVec,
+		podRequestsGaugeVec,
+		podLimitsGaugeVec,
+		daemonRequestsGaugeVec,
+		daemonLimitsGaugeVec,
+		overheadGaugeVec,
+	} {
+		gauge.Reset()
+	}
+}
+
+func (ns *nodeScraper) cleanup(ctx context.Context, existingNodes map[string]struct{}) {
+	nodesToRemove := []string{}
+	for nodeName := range ns.labelMap {
+		if _, ok := existingNodes[nodeName]; !ok {
+			nodesToRemove = append(nodesToRemove, nodeName)
+		}
+	}
+
+	// Remove all gauges associated with removed node
+	for _, node := range nodesToRemove {
+		gaugeMap := ns.labelMap[node]
+		for gaugeVec, labelSet := range gaugeMap {
+			for _, labels := range labelSet {
+				gaugeVec.Delete(labels)
+			}
+		}
+		delete(ns.labelMap, node)
+	}
+
+	if len(nodesToRemove) > 0 {
+		logging.FromContext(ctx).Infof("Removing the following node gauges: %s", func() string {
+			var sb strings.Builder
+			for idx, nodeName := range nodesToRemove {
+				sb.WriteString(nodeName)
+				if idx < len(nodesToRemove)-1 {
+					sb.WriteString(", ")
+				}
+			}
+			return sb.String()
+		}())
+	}
+}
+
+// set sets the value for the node gauge
+func (ns *nodeScraper) set(resourceList v1.ResourceList, node *v1.Node, gaugeVec *prometheus.GaugeVec) error {
+	for resourceName, quantity := range resourceList {
+		resourceTypeName := strings.ReplaceAll(strings.ToLower(string(resourceName)), "-", "_")
+		labels := ns.getNodeLabels(node, resourceTypeName)
+		ns.labelMap[node.Name][gaugeVec] = append(ns.labelMap[node.Name][gaugeVec], labels)
+
+		gauge, err := gaugeVec.GetMetricWith(labels)
+		if err != nil {
+			return fmt.Errorf("generate new gauge: %w", err)
+		}
+		if resourceName == v1.ResourceCPU {
+			gauge.Set(float64(quantity.MilliValue()) / float64(1000))
```

Comment:
Although this is the code that you're copying over, this looks a bit weird to do a MilliValue and then divide a 1000 again.